README.rst
openradar
=========

Radar
-----
This repository was made for a project to improve the rain monitoring for 
national precipitation in the Netherlands. The rain forecast and monitoring
was believed to thoroughly be improved with the use of more than just the
two Dutch radar stations:
    * De Bilt (Lat: 52.10168, Lon: 5.17834) and 
    * Den Helder (Lat: 52.95334, 4.79997) 
by using relevant radar stations of the neighbouring countries Belgium and Germany.

The project lives at http://nationaleregenradar.nl

The master script ``.venv/bin/master`` organizes and aggregates data (if necessary), 
both radar and gauge (ground stations) data. The products are delivered 
real-time, near-real-time and afterwards. Every product is delivered for
different time-steps: 5 minutes, 1 hour and 24 hours. 

Every 5 minutes data is collected from the different radars and gauge stations. 
Especially rain gauge data is not always delivered for that time interval. The
near-real-time data product thus should give more reliable data as more data
has arrived at that time. The aggregates (hourly and daily) are also used to 
calibrate the 5 minute data.


Development installation
------------------------

For development, you can use a docker-compose setup::

    $ docker-compose build --build-arg uid=`id -u` --build-arg gid=`id -g` lib
    $ docker-compose up --no-start
    $ docker-compose start
    $ docker-compose exec lib bash

(Re)create & activate a virtualenv::

    (docker)$ rm -rf .venv
    (docker)$ virtualenv --system-site-packages .venv
    (docker)$ source .venv/bin/activate

Install dependencies & package and run tests::

    (docker)(virtualenv)$ pip install -r requirements.txt
    (docker)(virtualenv)$ pip install -e .[test]
    (docker)(virtualenv)$ pytest

Update requirements.txt::
    
    (docker)$ rm -rf .venv
    (docker)$ virtualenv --system-site-packages .venv
    (docker)$ source .venv/bin/activate
    (docker)(virtualenv)$ pip install .
    (docker)(virtualenv)$ pip uninstall openradar --yes
    (docker)(virtualenv)$ pip freeze > requirements.txt


Server installation
-------------------

Global dependencies (apt)::

    git
    imagemagick
    libhdf5-serial-dev
    python3-gdal
    python3-pip
    python3-rpy2
    redis-server
    supervisor

Then, to install the 'gstat' package, in the R interpreter::
    
    > install.packages('gstat')

And give "y" when appropriate.

Then, the python part::

    $ sudo pip3 install --upgrade pip virtualenv
    $ virtualenv --system-site-packages .venv
    $ source .venv/bin/activate
    (virtualenv)$ pip install -r requirements.txt
    (virtualenv)$ pip install -e .


Checkout our private nens/rr-task in src/radar and use symbolic links to link to
some configuration files.

Make some symlinks:
    var/misc -> rr-task/misc/
    etc/supervisord.conf -> rr-task/<type>/supervisor/supervisord.conf
    openradar/localconfig -> rr-task/<type>/openradar/localconfig.py

To start the celery worker, either run `supervisord` in the installation
directory, or `supervisord -c /srv/openradar/etc/supervisord.conf` from
elsewhere.

Most scripts are periodically runned on the server via cronjobs. The respective
crontabs are (ideally) mirrored in our private nens/rr-task repository.


Scripts
-------
Scripts can be found in openradar/scripts

Scripts have an option --direct, to run without the task system.
Tasks have an argument --cascade. For most scripts this means creating
tasks for logical next steps. For the publish task, it means 'publish
any rescaled products as well.'

TODO: cover sync* scripts and partial scripts here, too.


Timezone
--------
Timezones:
- The time zones for all of the data is in UTC time.


Clutter filter
--------------
To update the clutter filter, execute this command::
    
    .venv/bin/clutter YYYYMMDD-YYYYMMDD -t ./my-clutter-file.h5

Put this file in the misc directory and update DECLUTTER_FILEPATH to
point to this file. The basename is enough, but an absolute path will
probably work, too.


Troubleshooting
---------------
The realtime products are a good indication for the times at which
master execution has not succesfully completed. To get a list of missing
products in the past 7 days run::

    $ .venv/bin/repair 7d

To get a hint about which masters to re-run.

Lately, there have been tasks hanging due to difficulties reaching or
writing to a configured share. In that case, try to stop celery, kill
any celery workers and start celery to see if the problem persists::

    $ supervisorctl shutdown

    Actions to kill remaining celery workers...

    $ supervisord

In extreme cases you could purge the task queue, but chances are that
the problem lies not in the tasks itself. It brings a lot of work to
resubmit the lost tasks. Anyway::

    $ .venv/bin/celery --app=openradar.tasks.app purge


Product table
-------------
This table shows how the products should be calibrated and which products
should be consistent with which other products::


    Timeframe | Product | Delivery*     | Calibration | Consistent with
    ----------+---------+---------------+-------------+----------------
              |    R    | Immediate     | Corr. Field |
    5 minutes |    N    | 1 hour        | Corr. Field | N - 1 hour
              |    A    | 12 hours      | Corr. Field | A - 1 hour
              |    U    | 30 days       | Corr. Field | U - 1 hour
    ----------+---------+---------------+-------------+----------------
              |    R    | Immediate     | Corr. Field |
     1 hour   |    N    | 1 hour        | Corr. Field |
              |    A    | 12 hours      | Kriging     | A - 1 day
              |    U    | 30 days       | Kriging     | U - 1 day
    ----------+---------+---------------+-------------+----------------
              |    R    | Immediate     | Corr. Field |
      1 day   |    N    | 1 hour        | Corr. Field |
              |    A    | 12 hours      | Kriging     |
              |    U    | 30 days       | Kriging     |


`*`: Delivery can not be earlier than the aggregated product that the
consistent product is based upon


CHANGES.rst
Changelog of openradar
===================================================


0.4.7 (unreleased)
------------------

- Used new clutterfile.

- Bumped radar to 0.3.8.

- Skip dotfiles in FTP sync source.

- Stop ftp publishing except for nowcast purposes.


0.4.6 (2018-12-21)
------------------

- Bugfix on previous png writing.


0.4.5 (2018-12-21)
------------------

- Write the previous png to reduce flickering.


0.4.4 (2018-11-09)
------------------

- Bumped radar to 0.3.7.


0.4.3 (2018-11-02)
------------------

- Used dwd asb.

- Used new clutterfile.

- Bumped radar to 0.3.6.


0.4.2 (2018-07-10)
------------------

- Created workaround for quirky opendata service.

- Added basic auth support and verification flag for remote radars.

- Add script to sync ground stations.

- Bumped radar to 0.3.5.

- Use new clutterfile.


0.4.1 (2017-12-29)
------------------

- Remove NL60.

- Adjust elevation index for NL61, because it is a new radar.

- Have NL61 and NL62 cover each others inner circles.

- Have JABs inner circle covered by NL62 instead of NL60.

- Add time_limit for all tasks.

- Work towards new HDF5 format.

- Put ciso8601 and dateparsing back for repair script.

- Resolve the numpy future warning.

- Remove automatic geotiff publishing.

- Set maximum retrieve time for volume data to 1 hour.

- Started using HTTP for fetching part of the volume data.

- Traded bootstrap.py for a pip-installed buildout.


0.4 (2016-08-26)
----------------

- Add Jabbeke.

- Add basic nowcast product, task and script

- Improve image script product selection

- Remove groundfile code, add grounddata from database code

- Improved the atomic report code

- Many improvements to atomic scripts for raster store management

- Remove catch-all style exception handling from tasks

- Celery broker url configurable in settings

- Fix atomic move stop date when store end comes before chunk end

- Update of raster store to fix problem with simultaneous merge and store


0.3.7 (2013-08-29)
------------------

- Fixed independent volume fetching when one of the providers fails.


0.3.6 (2013-08-29)
------------------

- Speedup sync_radar_to_ftp by caching remote listings.


0.3.5 (2013-08-29)
------------------

- Add sync_radar_to_ftp script to fill holes from ftp outages.


0.3.4 (2013-06-17)
------------------

- Add ftp to default arguments for publishing.


0.3.3 (2013-06-11)
------------------

- Change max_age for fetching files to one day.


0.3.2 (2013-06-06)
------------------

- Fix products being published for product 'a' timeframe 'h'


0.3.1 (2013-02-14)
------------------

- Adjust unit and coerce value to float for opendap data.


0.3 (2013-02-14)
----------------

- Change threddsfile time:unit attribute to time:units

- Add opendap retrieval functionality to ThreddsFile

- Add function to retrieve data from multiple threddsfiles.


0.2.10 (2013-02-13)
-------------------

- Fix the last SHAPE_DIR references.


0.2.9 (2013-02-13)
------------------

- Correct path to osm image and remove from png making.


0.2.8 (2013-02-13)
------------------

- Fix bug regarding missing shapedir.


0.2.7 (2013-02-13)
------------------

- Major rewrite of configuration system. Most configuration is now moved to this package. Stuff may be broken now, but it enables standalone testing of this library in the future.


0.2.6 (2013-02-12)
------------------

- Publish a merged threddsfile as well, where the realtime products
  are overwritten by near-realtime products, etc. The available variable
  functions as a flag to indicate what data was written.

- Range is now an optional positional argument.


0.2.5 (2013-01-25)
------------------

- Create the animated gif as tempfile, then move to actual target.


0.2.4 (2013-01-24)
------------------

- Add animated gif creation.


0.2.3 (2013-01-22)
------------------

- Add script for syncing of ftp folders (hirlam, eps)


0.2.2 (2013-01-21)
------------------

- Even less crashing of threddsfiles.


0.2.1 (2013-01-18)
------------------

- No logging for FtpImporter if nothing fetched.

- Fix crashing when creating new threddsfiles.


0.2.0 (2013-01-17)
------------------

- Format logging and show ftp result summary in logfile.

- No longer prepare google gtiff for web viewer.

- Delete existing h5 when creating, even if opening in 'w' mode.


0.1.9 (2013-01-17)
------------------

- Add ftp info to config

- Add FtpImporter that imports directly from scanfile sources.


0.1.8 (2013-01-17)
------------------

- CSV is now read from zipfile if it exists.

- Organize now moves anything with a csv extension to a zipped csv.


0.1.7 (2013-01-16)
------------------

- Separate error handling for creation and publishing.

- Update existing threddsfiles when publishing.

- Moving to much larger threddsfiles of around 10000 grids.


0.1.6 (2013-01-10)
------------------

- Re-enable multiprocessing.

- Fix n, h products not being published.

- Do local copying to subfolders per product.


0.1.5 (2013-01-10)
------------------

- Improved publishing routine. Will publish in order of importance.

- Created FtpPublisher that takes care of using folders on ftp.

- Tweaks to log messages.


0.1.4 (2013-01-09)
------------------

- Fix bug that tried to make d product at 9


0.1.3 (2013-01-09)
------------------

- Fix bug in get method of ConsistentProduct.


0.1.2 (2013-01-09)
------------------

- Set threshold for rain to 0.008 (that is 0.1 / 12)


0.1.1 (2013-01-09)
------------------

- Add master script.


0.1 (2013-01-09)
----------------

- Initial project structure created with nensskel 1.30.dev0.


openradar/utils.py
# -*- coding: utf-8 -*-
# (c) Nelen & Schuurmans.  GPL licensed, see LICENSE.rst.

from functools import reduce

import ciso8601
import codecs
import datetime
import h5py
import logging
import numpy as np
import os

from osgeo import osr
from matplotlib import colors
from matplotlib import cm
from PIL import Image

from openradar import config

# Some projections
UTM = 3405
DHDN = 4314
WGS84 = 4326
GOOGLE = 900913

AEQD_PROJ4 = ('+proj=aeqd +a=6378.137 +b=6356.752 +R_A'
              ' +lat_0={lat} +lon_0={lon} +x_0=0 +y_0=0')

# Copied from lizard_map/coordinates.py
RD = ("+proj=sterea +lat_0=52.15616055555555 +lon_0=5.38763888888889 "
      "+k=0.999908 +x_0=155000 +y_0=463000 +ellps=bessel "
      "+towgs84=565.237,50.0087,465.658,-0.406857,0.350733,-1.87035,4.0812 "
      "+units=m +no_defs")


# Projections and transformations
def projection_aeqd(lat=None, lon=None):
    sr = osr.SpatialReference()
    sr.ImportFromProj4(str(AEQD_PROJ4.format(lat=lat, lon=lon)))
    return sr.ExportToWkt()


def projection(desc, export='wkt'):
    sr = osr.SpatialReference()
    if isinstance(desc, int):
        sr.ImportFromEPSG(desc)
    elif isinstance(desc, (str)):
        if desc.startswith('+proj='):
            sr.ImportFromProj4(str(desc))
        else:
            sr.ImportFromWkt(str(desc))
    if export == 'wkt':
        return sr.ExportToWkt()
    if export == 'proj4':
        return sr.ExportToProj4()
    return sr


def transform(point, desc):
    srs = [projection(e, export=None) for e in desc]
    ct = osr.CoordinateTransformation(*srs)
    return ct.TransformPoint(*point)[0:2]


class CoordinateTransformer(object):
    """
    Transform coordinates or give cached coordinates back.

    TODO Make this a simple function, caching is not trivial here.
    """

    def __init__(self):
        self.cache = {}

    def transform(self, points, projections):
        """
        Transform arrays of points from one projection to another.
        """
        shape = np.array([p.shape for p in points]).max(0)

        points_in = np.array([
            points[0].flatten(),
            points[1].flatten(),
        ]).transpose()

        ct = osr.CoordinateTransformation(
            projection(projections[0], export=None),
            projection(projections[1], export=None),
        )

        points_out = np.array(ct.TransformPoints(points_in))[:, 0:2]

        result = points_out.reshape(shape[0], shape[1], 2).transpose(2, 0, 1)
        return result


# Singleton
coordinate_transformer = CoordinateTransformer()


# Dateranges and date helpers
def datetime_range(start, stop, step):
    """ Return generator of datetimes. """
    datetime = start
    while datetime <= stop:
        yield datetime
        datetime += step


def variabletimestamp2datetime(ts, fmt='%Y%m%d%H%M%S'):
    """ Trying to match and increasingly detailed timestamp. """
    return datetime.datetime.strptime(ts, fmt[:len(ts) - 2])


def timestamp2datetime(ts, fmt='%Y%m%d%H%M%S'):
    return datetime.datetime.strptime(ts, fmt)


def datetime2timestamp(dt, fmt='%Y%m%d%H%M%S'):
    return dt.strftime(fmt)


class DateRange(object):

    STEP = datetime.timedelta(minutes=5)

    def __init__(self, text, step=STEP):
        self.step = step
        self.start, self.stop = self._start_stop_from_text(text)

    def _start_stop_from_text(self, text):
        """
        Return start and end timestamps.
        """
        if '-' in text:
            text1, text2 = text.split('-')
            start = self._single_from_text(text=text1)
            stop = self._single_from_text(text=text2, last=True)
        else:
            start = self._single_from_text(text=text)
            stop = self._single_from_text(text=text)
        return start, stop

    def _single_from_text(self, text, last=False):
        """
        Return datetime that matches text.

        If last, return the last possible step for text.
        """
        first = variabletimestamp2datetime(text)
        if not last:
            return first

        td_kwargs = {
            8: {'days': 1},
            10: {'hours': 1},
            12: {'minutes': 5},  # Makes a range of minutes possible.
        }[len(text)]
        return first - self.step + datetime.timedelta(**td_kwargs)

    def iterdatetimes(self):
        dt = self.start
        while dt <= self.stop:
            yield dt
            dt += self.step


class MultiDateRange(object):
    """ Hold a series of datetime ranges and generate it. """

    def __init__(self, text, step=5):
        step = datetime.timedelta(minutes=step)
        self._dateranges = [DateRange(subtext, step)
                            for subtext in text.strip(',').split(',')]

    def iterdatetimes(self):
        for dr in self._dateranges:
            for dt in dr.iterdatetimes():
                yield dt


# Path helpers and organizers
class PathHelper(object):
    """
        >>> import datetime
        >>> dt = datetime.datetime(2011, 03, 05, 14, 15, 00)
        >>> ph = PathHelper('a', 'c', '{code}:{timestamp}')
        >>> ph. path(dt)
        u'a/c/2011/03/05/c:20110305141500'
    """

    TIMESTAMP_FORMAT = config.TIMESTAMP_FORMAT

    def __init__(self, basedir, code, template):
        """
        Filetemplate must be something like '{code}_{timestamp}.'
        """
        self._basedir = os.path.join(basedir, code)
        self._code = code
        self._template = template

    def path(self, dt):
        return os.path.join(
            self._basedir,
            dt.strftime('%Y'),
            dt.strftime('%m'),
            dt.strftime('%d'),
            self._template.format(
                code=self._code,
                timestamp=dt.strftime(self.TIMESTAMP_FORMAT)
            )
        )

    def path_with_hour(self, dt):
        return os.path.join(
            self._basedir,
            dt.strftime('%Y'),
            dt.strftime('%m'),
            dt.strftime('%d'),
            dt.strftime('%H'),
            self._template.format(
                code=self._code,
                timestamp=dt.strftime(self.TIMESTAMP_FORMAT)
            )
        )


# Timing
def closest_time(timeframe='f', dt_close=None):
    '''
    Get corresponding datetime based on the timeframe.
    e.g. with    dt_close = 2012-04-13 12:27
    timeframe = 'f' returns 2012-04-13 12:25
    timeframe = 'h' returns 2012-04-13 12:00
    timeframe = 'd' returns 2012-04-12 08:00
    '''
    if dt_close is not None:
        now = dt_close
    else:
        now = datetime.datetime.utcnow()
    if timeframe == 'h':
        closesttime = now.replace(minute=0, second=0, microsecond=0)
    elif timeframe == 'd':
        closesttime = now.replace(hour=8, minute=0, second=0, microsecond=0)
        if closesttime > now:
            closesttime = closesttime - datetime.timedelta(days=1)
    else:
        closesttime = now.replace(
            minute=now.minute - (now.minute % 5), second=0, microsecond=0,
        )
    return closesttime


def get_valid_timeframes(datetime):
    """ Return a list of timeframe codes corresponding to a datetime."""
    result = []
    if datetime.second == 0 and datetime.microsecond == 0:
        if datetime.minute == (datetime.minute // 5) * 5:
            result.append('f')
        if datetime.minute == 0:
            result.append('h')
            if datetime.hour == 8:
                result.append('d')
    return result


def get_aggregate_combinations(datetimes,
                               timeframes=['f', 'h', 'd']):
    """ Return generator of dictionaries. """
    for _datetime in datetimes:
        valid_timeframes = get_valid_timeframes(datetime=_datetime)
        for timeframe in timeframes:
            if timeframe in valid_timeframes:
                yield dict(nowcast=False,
                           datetime=_datetime,
                           timeframe=timeframe)
                if timeframe == 'f':
                    yield dict(nowcast=True,
                               datetime=_datetime,
                               timeframe=timeframe)


def get_product_combinations(datetimes,
                             prodcodes=['r', 'n', 'a', 'u'],
                             timeframes=['f', 'h', 'd']):
    """ Return generator of dictionaries. """
    for _datetime in datetimes:
        valid_timeframes = get_valid_timeframes(datetime=_datetime)
        for prodcode in prodcodes:
            for timeframe in timeframes:
                if timeframe in valid_timeframes:
                    yield dict(nowcast=False,
                               datetime=_datetime,
                               prodcode=prodcode,
                               timeframe=timeframe)
                    if timeframe == 'f' and prodcode == 'r':
                        yield dict(nowcast=True,
                                   datetime=_datetime,
                                   prodcode=prodcode,
                                   timeframe=timeframe)


def consistent_product_expected(prodcode, timeframe):
    """
    Return if a consistent product is expected, for a product. It
    can be used to determine if a product needs to be published, or
    not. If not, the rescaled equivalent should be published.
    """
    if prodcode in 'au' and timeframe in 'fh':
        return True
    if prodcode == 'n' and timeframe == 'f':
        return True
    return False


def get_groundfile_datetimes(prodcode, timeframe, date):
    '''
    Return datetime for groundfile for a given product code and datetime

    For ground data to be more complete, the datetime of the grounddata
    must a certain amount later than the radar datetime.  So for a
    product at 2012-12-18 09:05 the groundfile datetimes must be:
        real-time           => 2012-12-18-09:05
        near-real-time      => 2012-12-18-10:05
        afterwards          => 2012-12-20-09:05

    Unfortunately, for technical reasons, the groundfile dating is
    not consistent over history. Therefore, this function now takes a
    timeframe argument and returns a generator of datetimes to try. The
    first datetime for which a ground file exists, must be used.
    '''
    delta = dict(
        r=datetime.timedelta(minutes=0),
        n=datetime.timedelta(hours=1),
        a=datetime.timedelta(days=2)
    )
    base_datetime = date + delta[prodcode]
    if timeframe == 'f':
        # It used to be on exact 5 minutes, but now it is 1 minute late.
        for minutes in [1, 0]:
            yield base_datetime + datetime.timedelta(minutes=minutes)
    elif timeframe == 'd' and prodcode == 'n':
        # It used to be on 8 o'clock utc, but now it is (correctly) at 9.
        for hours in [0, -1]:
            yield base_datetime + datetime.timedelta(hours=hours)
    else:
        yield base_datetime


# Visualization
def rain_kwargs(max_rain=120, name='buienradar', threshold=0.1):
    """ Return colormap and normalizer suitable for rain. """
    if name == 'buienradar':
        rain_colors = {
            'red': (
                (0, 240, 240),
                (2, 158, 110),
                (5, 88, 0),
                (10, 0, 255),
                (100, 131, 192),
                (max_rain, 192, 192),
            ),
            'green': (
                (0, 240, 240),
                (2, 158, 110),
                (5, 88, 0),
                (10, 0, 0),
                (100, 0, 0),
                (max_rain, 0, 0),
            ),
            'blue': (
                (0, 255, 255),
                (2, 255, 255),
                (5, 255, 200),
                (10, 110, 0),
                (100, 0, 192),
                (max_rain, 192, 192),
            ),
        }

        cdict = {}
        for key, value in rain_colors.items():
            cdict[key] = []
            for element in value:
                cdict[key].append((
                    element[0] / max_rain,
                    element[1] / 255,
                    element[2] / 255,
                ))

        colormap = colors.LinearSegmentedColormap('rain', cdict)
        normalize = colors.Normalize(vmin=0, vmax=max_rain)

        return dict(colormap=colormap, normalize=normalize)

    if name == 'jet':
        colormap = cm.jet

        def normalize(data):
            ma = np.ma.array(data)
            ma[np.less(ma, threshold)] = np.ma.masked
            return colors.LogNorm(vmin=threshold, vmax=max_rain)(ma)

        return dict(colormap=colormap, normalize=normalize)


def merge(images):
    """
    Return a pil image.

    Merge a list of pil images with equal sizes top down based on
    the alpha channel.
    """

    def paste(image1, image2):
        image = image2.copy()
        mask = Image.fromarray(np.array(image1)[:, :, 3])
        rgb = Image.fromarray(np.array(image1)[:, :, 0:3])
        image.paste(rgb, None, mask)
        return image

    return reduce(paste, images)


def makedir(dirname):
    """ Return True if directory was created, else False. """
    try:
        os.makedirs(dirname)
        return True
    except OSError:
        return False


class UTF8Recoder:
    """
    Iterator that reads an encoded stream and reencodes the input to UTF-8
    """
    def __init__(self, f, encoding):
        self.reader = codecs.getreader(encoding)(f)

    def __iter__(self):
        return self

    def next(self):
        return self.reader.next().encode("utf-8")


def save_attrs(h5, attrs):
    for key, value in attrs.items():
        if isinstance(value, dict):
            group = h5.create_group(key)
            save_attrs(h5=group, attrs=value)
            continue
        if hasattr(value, 'encode'):
            value = value.encode('ascii')
        h5.attrs[key] = np.array([value])


def save_dataset(data, meta, path):
    '''
    Accepts an array jampacked with data, a metadata file and a path
    to produce a fresh h5 file.
    '''
    logging.debug('Saving hdf5 dataset: {}'.format(os.path.basename(path)))
    logging.debug(path)
    tmp_path = path + '.in'

    makedir(os.path.dirname(path))
    h5 = h5py.File(tmp_path, 'w')

    # Geographic group
    left, right, top, bottom = meta['grid_extent']
    width, height = meta['grid_size']

    geographic = dict(
        geo_par_pixel=b'X,Y',
        geo_dim_pixel=b'KM,KM',
        geo_pixel_def=b'CENTRE',
        geo_number_columns=width,
        geo_number_rows=height,
        geo_pixel_size_x=1.000,
        geo_pixel_size_y=-1.000,
        geo_product_corners=[left, bottom, left, top,
                             right, top, right, bottom],
        map_projection=dict(
            projection_indication=b'Y',
            projection_name=b'STEREOGRAPHIC',
            projection_proj4_params=projection(RD, export='proj4'),
        ),
    )

    # Overview group
    availables = meta['available']
    try:
        availables_any = [any(r) for r in zip(*availables)]
    except TypeError:
        availables_any = availables
    products_missing = str(', '.join(
        [radar
         for radar, available in zip(meta['radars'], availables_any)
         if not available],
    ))
    timestamp_last_composite = meta['timestamp_last_composite']
    if hasattr(timestamp_last_composite, "decode"):
        timestamp_last_composite = timestamp_last_composite.decode("ascii")
    product_datetime_start = (timestamp2datetime(
        timestamp_last_composite,
    ) + config.TIMEFRAME_DELTA['f']).strftime('%d-%b-%Y;%H:%M:%S.%f').upper()
    product_datetime_end = product_datetime_start

    overview = dict(
        hdftag_version_number=b'3.5',
        number_image_groups=1,
        number_radar_groups=0,
        number_satellite_groups=0,
        number_station_groups=0,
        product_datetime_end=product_datetime_end,
        product_datetime_start=product_datetime_start,
        product_group_name=str(os.path.splitext(os.path.basename(path))[0]),
        products_missing=products_missing,
    )

    # Image group
    calibration = dict(
        calibration_flag=b'Y',
        calibration_formulas=b'GEO = 0.010000 * PV + 0.000000',
        calibration_missing_data=0,
        calibration_out_of_image=65535,
    )

    image1 = dict(
        calibration=calibration,
        image_bytes_per_pixel=2,
        image_geo_parameter=b'PRECIP_[MM]',
        image_product_name=overview['product_group_name'],
        image_size=data['precipitation'].size,
    )

    groups = dict(
        overview=overview,
        geographic=geographic,
        image1=image1,
    )

    save_attrs(h5, groups)
    dataset = h5.create_dataset('image1/image_data',
                                data['precipitation'].shape,
                                dtype='u2', compression='gzip', shuffle=True)

    image_data = dict(
        CLASS=b'IMAGE',
        VERSION=b'1.2',
    )

    save_attrs(dataset, image_data)

    # Creating the pixel values
    dataset[...] = np.uint16(np.round(data['precipitation'] * 100)).filled(
        calibration['calibration_out_of_image'],
    )

    # Keep the old way for compatibility with various products
    for name, value in data.items():
        dataset = h5.create_dataset(name, value.shape, dtype='f4',
                                    compression='gzip', shuffle=True)
        dataset[...] = value.filled(config.NODATAVALUE)

    for name, value in meta.items():
        if hasattr(value, 'encode'):
            value = value.encode('ascii')
        elif isinstance(value, list):
            value = [
                e.encode('ascii')
                if hasattr(e, 'encode')
                else e
                for e in value
            ]
        h5.attrs[name] = value

    h5.close()
    os.rename(tmp_path, path)

    return path


def get_countrymask():
    """
    Get a prefabricated country mask, 1 everywhere in the country and 0
    50 km outside the country. If extent and cellsize are not as in
    config.py, this breaks.
    """
    countrymask_path = os.path.join(config.MISC_DIR, 'countrymask.h5')
    h5 = h5py.File(countrymask_path, 'r')
    mask = h5['mask'][...]
    h5.close()
    return mask


def get_geo_transform():
    left, right, top, bottom = config.COMPOSITE_EXTENT
    width, height = config.COMPOSITE_CELLSIZE
    return left, width, 0, top, 0, -height


def parse_datetime(text):
    """ Return a datetime instance. """
    if len(text) == 4:
        # ciso8601 requires at least months to be specified
        return ciso8601.parse_datetime_unaware(text + '01')
    return ciso8601.parse_datetime_unaware(text)


def convert_to_lists_and_unicode(meta):
    """ Updates dict. Return None. """
    for k, v in meta.items():
        # numpy array
        if hasattr(v, 'size'):
            if v.size == 0:
                meta[k] = []
            elif v.size == 1:
                meta[k] = v.item()
            else:
                meta[k] = [
                    e.decode('ascii')
                    if hasattr(e, 'decode') else e
                    for e in v.tolist()
                ]
        # bytes
        elif hasattr(v, 'decode'):
            meta[k] = v.decode('ascii')
